File: bindgen-tests/ruby/tests/bytes.rb
# frozen_string_literal: true

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

require 'test/unit'
require 'uniffi_bindgen_tests'

class TestBytes < Test::Unit::TestCase
  include UniffiBindgenTests

  def test_roundtrip_bytes
    data = 'test-data'.b

    assert_equal data, UniffiBindgenTests.roundtrip_bytes(data)
  end

  def test_roundtrip_empty_bytes
    assert_equal ''.b, UniffiBindgenTests.roundtrip_bytes(''.b)
  end

  def test_roundtrip_binary_bytes
    data = "\x00\x01\x02\xFF".b

    assert_equal data, UniffiBindgenTests.roundtrip_bytes(data)
  end

  def test_bytes_encoding
    result = UniffiBindgenTests.roundtrip_bytes('hello'.b)

    assert_equal Encoding::BINARY, result.encoding
  end

  # Zero-copy &[u8] — proc-macro path
  def zero_copy_bytes
    assert_equal 0, sum_bytes_procmacro(''.b)
    assert_equal 6, sum_bytes_procmacro("\x01\x02\x03".b)
    assert_nil first_byte_procmacro(''.b)
    assert_equal 42, first_byte_procmacro("\x2a".b)
  end
end

File: bindgen-tests/swift/tests/bytes.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import uniffi_bindgen_tests
import Foundation

var data = Data.init()
data.append(contentsOf: [1, 2, 3, 4])
assert(roundtripBytes(a: data) == data);

// Zero-copy &[u8]
assert(sumBytes(buf: Data()) == 0)
assert(sumBytes(buf: Data([1, 2, 3])) == 6)
assert(firstByte(buf: Data()) == nil)
assert(firstByte(buf: Data([42])) == 42)

// Zero-copy &mut [u8]. Rust writes land in the caller's Data.
var fillMe = Data(count: 4)
fillBytes(buf: &fillMe)
assert(fillMe == Data([0, 1, 2, 3]))

var incMe = Data([1, 2, 3])
incrementBytes(buf: &incMe)
assert(incMe == Data([2, 3, 4]))

// Empty buffer must not crash.
var empty = Data()
fillBytes(buf: &empty)
assert(empty == Data())
